Sources/SwiftSyntaxMacrosGenericTestSupport/Assertions.swift

```diff
@@ -338,9 +338,36 @@ extension DiagnosticSpec {
   }
 }
 
-func assertDiagnostic(
+/// Describes the context in which we are asserting diagnostic correctness.
+///
+/// This is used to map source locations.
+public enum DiagnosticAssertionContext {
+  case macroExpansion(BasicMacroExpansionContext)
+  case tree(any SyntaxProtocol)
+
+  func location(
+    for position: AbsolutePosition,
+    anchoredAt node: Syntax,
+    fileName: String
+  ) -> SourceLocation {
+    switch self {
+    case .macroExpansion(let expansionContext):
+      return expansionContext.location(
+        for: position,
+        anchoredAt: node,
+        fileName: fileName
+      )
+
+    case .tree(let syntax):
+      return SourceLocationConverter(fileName: fileName, tree: syntax)
+        .location(for: position)
+    }
+  }
+}
+
+public func assertDiagnostic(
```

Comment:
Did you intentionally make this API public? I would prefer to not expose it. If you just want to use it in `SwiftIfConfigTest` we should mark it as `@_spi(Testing)`.

Reply:
Yes, I need it. I'll mark it `@_spi`.